### PR TITLE
Sprint1

### PR DESCRIPTION
**Pull request by Tratori**

### CODE REVIEW

File: src/lib/storage/table.cpp [hunk 1]
```diff
@@ -1,84 +1,108 @@
...
+    throw std::logic_error("Table does not contain chunk with the requested id.");
+  }
+  return _chunks[chunk_id];
 }
 
 std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _chunks[chunk_id];
 }
```
Comment (Tratori) [inline src/lib/storage/table.cpp]: Is the upper method supposed to return a copy, perhaps?
why else would we distinguish between const and not const?

File: src/lib/storage/chunk.cpp [hunk 2]
```diff
@@ -1,31 +1,65 @@
 #include "chunk.hpp"
 
-#include "abstract_segment.hpp"
 #include "utils/assert.hpp"
 
 namespace opossum {
 
-void Chunk::add_segment(const std::shared_ptr<AbstractSegment> segment) {
+void Chunk::add_segment(const std::shared_ptr<AbstractSegment> column) {
   // Implementation goes here
```
Comment (Tratori) [inline src/lib/storage/chunk.cpp]: Not sure whether we should keep the original signature (not rename segment to column). If we want to stick with column we should also rename it in the .hpp though.

We should also probably remove the comments stating "Implementation goes here"

File: src/lib/storage/storage_manager.cpp [hunk 3]
```diff
@@ -5,42 +5,53 @@
...
-  return *(new StorageManager());
-  // A really hacky fix to get the tests to run - replace this with your implementation
+  static StorageManager instance;
+  return instance;
 }
 
 void StorageManager::add_table(const std::string& name, std::shared_ptr<Table> table) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  _tables[name] = table;
```
Comment (Tratori) [inline src/lib/storage/storage_manager.cpp]: check if table already exists

File: src/lib/storage/storage_manager.hpp [hunk 4]
```diff
@@ -38,7 +38,7 @@ class StorageManager : private Noncopyable {
   StorageManager() {}
 
   StorageManager& operator=(StorageManager&&) = default;
-
+  std::map<std::string, std::shared_ptr<Table>> _tables;
```
Comment (Tratori) [inline src/lib/storage/storage_manager.hpp]: unordered_map should be preferred

File: src/lib/storage/storage_manager.cpp (hunk 3 above)
Comment (Tratori) [reply src/lib/storage/storage_manager.cpp]: might also want to add a test for that :D

File: src/lib/storage/chunk.cpp [hunk 5]
```diff
@@ -1,31 +1,63 @@
...
 }
 
 void Chunk::append(const std::vector<AllTypeVariant>& values) {
   // Implementation goes here
+  DebugAssert(values.size() == column_count(), "Cannot insert a tuple with less values than columns.");
+
+  auto column_it = _columns.begin();
+  auto value_it = values.begin();
+
+  while (column_it != _columns.end()) {
```
Comment (rothaarlappen) [inline src/lib/storage/chunk.cpp]: Why do we use a while-loop here? Wouldn't a for-loop be sufficient as well, while saving 3 LoC? ^^

File: src/lib/storage/chunk.cpp [hunk 6]
```diff
@@ -1,31 +1,63 @@
...
+    // Problem: How to iterate over the given data types in all_type_variant?
+    const auto no_types = types.storage_.size_;
+    DebugAssert(no_types == 5, "There are 5 supported data types but a different count was found.");
+
+    append_with_type<int32_t>(column_it, value_it);
+    append_with_type<int64_t>(column_it, value_it);
+    append_with_type<float>(column_it, value_it);
+    append_with_type<double>(column_it, value_it);
+    append_with_type<std::string>(column_it, value_it);
```
Comment (rothaarlappen) [inline src/lib/storage/chunk.cpp]: Only one of those should succeed and return "true"...  
What about rewriting into 
`if(append_with_type<int32_t>(column_it, value_it) || 
append_with_type<int64_t>(column_it, value_it)) || ...`

or 
'if () {}
else if () {} 
...' 

to stop evaluating after "correct datatype" is found? (Would also allow handling cases were typecasting doesn't succeed at all...)

File: src/lib/storage/table.hpp [hunk 7]
```diff
@@ -67,6 +67,11 @@ class Table : private Noncopyable {
   void compress_chunk(const ChunkID chunk_id);
 
  protected:
+  std::vector<std::shared_ptr<Chunk>> _chunks;
+  std::vector<std::string> _column_names;
+  std::vector<std::string> _column_types;
+  std::vector<bool> _column_nullable;
```
Comment (rothaarlappen) [inline src/lib/storage/table.hpp:73]: should this be renamed into "_column_nullables"?

File: src/lib/storage/value_segment.cpp [hunk 8]
```diff
@@ -7,67 +7,90 @@ namespace opossum {
...
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _values;
 }
 
 template <typename T>
 bool ValueSegment<T>::is_nullable() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return static_cast<bool>(_null_values);
```
Comment (rothaarlappen) [inline src/lib/storage/value_segment.cpp]: I would prefer 

`return _null_values.has_value()`

File: src/lib/storage/value_segment.cpp [hunk 9]
```diff
@@ -7,67 +7,90 @@ namespace opossum {
...
 }
 
 template <typename T>
 const std::vector<bool>& ValueSegment<T>::null_values() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (!is_nullable()) {
+    throw std::logic_error("Segment is not nullable.");
+  }
+  return *_null_values;
```
Comment (rothaarlappen) [inline src/lib/storage/value_segment.cpp:88]: Maybe 
`return _null_values.value()`

File: src/lib/storage/value_segment.cpp [hunk 10]
```diff
@@ -7,67 +7,90 @@ namespace opossum {
...
+    _null_values = std::vector<bool>();
+  }
 }
 
 template <typename T>
 AllTypeVariant ValueSegment<T>::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  DebugAssert(chunk_offset < size(), "Out of bounds.");
+  if (is_nullable() && (*_null_values)[chunk_offset]) {
```
Comment (rothaarlappen) [inline src/lib/storage/value_segment.cpp]: I think we should use the is_null method in the if-condition

File: src/lib/storage/value_segment.cpp [hunk 11]
```diff
@@ -7,67 +7,90 @@ namespace opossum {
...
+  DebugAssert(chunk_offset < size(), "Out of bounds.");
+  return is_nullable() && (*_null_values)[chunk_offset];
 }
 
 template <typename T>
 T ValueSegment<T>::get(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  DebugAssert(chunk_offset < size(), "Out of bounds.");
+  if (is_nullable() && (*_null_values)[chunk_offset]) {
```
Comment (rothaarlappen) [inline src/lib/storage/value_segment.cpp]: We could reuse is_null method for this one as well :D

File: src/lib/storage/value_segment.cpp [hunk 12]
```diff
@@ -7,67 +7,90 @@ namespace opossum {
...
+  }
+  return _values[chunk_offset];
 }
 
 template <typename T>
 std::optional<T> ValueSegment<T>::get_typed_value(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  DebugAssert(chunk_offset < size(), "Out of bounds.");
+  if (is_nullable() && (*_null_values)[chunk_offset]) {
```
Comment (rothaarlappen) [inline src/lib/storage/value_segment.cpp]: Another one.

File: src/lib/storage/value_segment.cpp [hunk 13]
```diff
@@ -7,67 +7,90 @@ namespace opossum {
...
 
 template <typename T>
 void ValueSegment<T>::append(const AllTypeVariant& value) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (variant_is_null(value)) {
+    if (!is_nullable()) {
+      throw std::logic_error("Trying to append NULL value to non nullable segment.");
+    }
+    (*_null_values).push_back(variant_is_null(value));
```
Comment (rothaarlappen) [inline src/lib/storage/value_segment.cpp]: variant_is_null(value) is always true in this branch

File: src/lib/storage/chunk.hpp [hunk 14]
```diff
@@ -37,6 +36,8 @@ class Chunk : private Noncopyable {
 
  protected:
   // Implementation goes here
```
Comment (rothaarlappen) [inline src/lib/storage/chunk.hpp]: Remove this comment

File: src/lib/storage/chunk.cpp [hunk 15]
```diff
@@ -1,31 +1,63 @@
...
+
+    ++column_it;
+    ++value_it;
+  }
 }
 
 std::shared_ptr<AbstractSegment> Chunk::get_segment(const ColumnID column_id) const {
   // Implementation goes here
```
Comment (rothaarlappen) [inline src/lib/storage/chunk.cpp]: remove comment

**Review by rothaarlappen — CHANGES_REQUESTED** [review]

File: src/lib/storage/chunk.cpp (hunk 2 above)
Comment (judgeNotFound) [reply src/lib/storage/chunk.cpp]: Lets change as few as possible.

File: src/lib/storage/chunk.cpp (hunk 5 above)
Comment (judgeNotFound) [reply src/lib/storage/chunk.cpp]: I guess for iterators, while loops are more common, but it is pretty much irrelevant.

File: src/lib/storage/chunk.cpp (hunk 6 above)
Comment (judgeNotFound) [reply src/lib/storage/chunk.cpp]: I thought about the second alternative too, but did not really like it. The first one is nice though.

File: src/lib/storage/value_segment.cpp (hunk 9 above)
Comment (Tratori) [reply src/lib/storage/value_segment.cpp:88]: Why?

File: src/lib/storage/table.cpp (hunk 1 above)
Comment (rothaarlappen) [reply src/lib/storage/table.cpp]: Chunk is a subclass of Noncopyable... Actually, I don't know how those two methods should differ...

File: src/lib/storage/table.cpp [hunk 16]
```diff
@@ -1,84 +1,111 @@
...
 ColumnCount Table::column_count() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return static_cast<ColumnCount>(_column_names.size());
 }
 
 uint64_t Table::row_count() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return (_chunks.size() - 1) * _target_chunk_size + _chunks.back()->size();
```
Comment (judgeNotFound) [inline src/lib/storage/table.cpp]: This works, when we are going to have immutable chunks only, but I am not sure about this. Probably a more simple approach having less assumptions:

```
auto sum = uint32_t{0};
for (const auto& chunk : _chunks) {
  sum += chunk.size();
}
```

File: src/lib/storage/table.cpp [hunk 17]
```diff
@@ -1,84 +1,111 @@
...
 ChunkID Table::chunk_count() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return static_cast<ChunkID>(_chunks.size());
 }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  for (unsigned int i = 0; i < _column_names.size(); i++) {
```
Comment (judgeNotFound) [inline src/lib/storage/table.cpp]: The chair likes `++i` instead of `i++`

File: src/lib/storage/table.cpp [hunk 18]
```diff
@@ -1,84 +1,111 @@
...
+    if (_column_names[i] == column_name) {
+      return static_cast<ColumnID>(i);
+    }
+  }
+  throw std::logic_error("Table does not contain column with the requested name.");
```
Comment (judgeNotFound) [inline src/lib/storage/table.cpp:78]: Might be easier to use in the future with smth like 

```
const auto row_id = table.column_id_by_name("columnX");
if (row_id) {
  ...
} else {
  // No row id found
}
```

```suggestion
  return null;
```

File: src/lib/storage/table.cpp [hunk 19]
```diff
@@ -1,84 +1,111 @@
...
-  Fail("Implementation is missing.");
+  return _column_names;
 }
 
 const std::string& Table::column_name(const ColumnID column_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (column_id > _column_names.size()) {
+    throw std::logic_error("Table does not contain column with the requested id.");
+  }
```
Comment (judgeNotFound) [inline src/lib/storage/table.cpp:92]:
```suggestion
    DebugAssert(column_id < _column_names.size(), "Table does not contain column with the requested id.");
```

File: src/lib/storage/table.cpp [hunk 20]
```diff
@@ -1,84 +1,111 @@
...
+  return _column_names[column_id];
 }
 
 const std::string& Table::column_type(const ColumnID column_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (column_id > _column_types.size()) {
+    throw std::logic_error("Table does not contain column with the requested id.");
+  }
```
Comment (judgeNotFound) [inline src/lib/storage/table.cpp:99]:
```suggestion
    DebugAssert(column_id < _column_names.size(), "Table does not contain column with the requested id.");
```

File: src/lib/storage/table.cpp [hunk 21]
```diff
@@ -1,84 +1,111 @@
...
+  return _column_types[column_id];
 }
 
 bool Table::column_nullable(const ColumnID column_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (column_id > _column_nullable.size()) {
+    throw std::logic_error("Table does not contain column with the requested id.");
+  }
```
Comment (judgeNotFound) [inline src/lib/storage/table.cpp:106]:
```suggestion
    DebugAssert(column_id < _column_names.size(), "Table does not contain column with the requested id.");
```

File: src/lib/storage/table.cpp [hunk 22]
```diff
@@ -1,84 +1,111 @@
...
+  return _column_nullable[column_id];
 }
 
 std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (chunk_id >= static_cast<ChunkID>(_chunks.size())) {
+    throw std::logic_error("Table does not contain chunk with the requested id.");
+  }
```
Comment (judgeNotFound) [inline src/lib/storage/table.cpp:113]:
```suggestion
    DebugAssert(chunk_id < static_cast<ChunkID>(_chunks.size()), "Table does not contain column with the requested id.");
```

File: src/lib/storage/table.cpp [hunk 23]
```diff
@@ -1,84 +1,111 @@
...
+  return _chunks[chunk_id];
 }
 
 std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (chunk_id >= static_cast<ChunkID>(_chunks.size())) {
+    throw std::logic_error("Table does not contain chunk with the requested id.");
+  }
```
Comment (judgeNotFound) [inline src/lib/storage/table.cpp:120]:
```suggestion
    DebugAssert(chunk_id < static_cast<ChunkID>(_chunks.size()), "Table does not contain column with the requested id.");
```

File: src/lib/storage/value_segment.cpp (hunk 9 above)
Comment (rothaarlappen) [reply src/lib/storage/value_segment.cpp:88]: I think it's easier to read but we can keep the *

File: src/lib/storage/table.cpp (hunk 16 above)
Comment (rothaarlappen) [reply src/lib/storage/table.cpp]: The exercise states that 

"When a table is created, an optioanle parameter defines the maximum size of a chunk. [...]. The maximum chunk size is stored in the table and cannot be changed. Inserts are always done into the last chunk, checking if this chunk has already reached its maximum size. If this is the case, a new chunk is created." 

What do you think? Should we change the implementation or keep it?

File: src/lib/storage/table.cpp (hunk 18 above)
Comment (rothaarlappen) [reply src/lib/storage/table.cpp:78]: The provided code included a test that expects a logic error if we lookup a columnid by name: 

```
TEST_F(StorageTableTest, GetColumnIdByName) {
  EXPECT_EQ(table.column_id_by_name("col_2"), 1);
  EXPECT_THROW(table.column_id_by_name("no_column_name"), std::logic_error);
}
```

File: src/lib/storage/table.cpp (hunk 19 above)
Comment (rothaarlappen) [reply src/lib/storage/table.cpp:92]: The tests expect a logic_error to be thrown here, otherwise I would agree.

**Pull request closed without merging**